File: 0673-number-of-longest-increasing-subsequence/0673-number-of-longest-increasing-subsequence.java
class Solution {
    public int findNumberOfLIS(int[] nums) {
        int n = nums.length, maxi = 1;
        int dp[] = new int[n];
        int cnt[] = new int[n];
        Arrays.fill(dp, 1);
        Arrays.fill(cnt, 1);
        
        for (int i = 1; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (nums[i] > nums[j] && dp[j]+1 > dp[i]) {
                    dp[i] = dp[j]+1;
                    cnt[i] = cnt[j];
                } else if (nums[i] > nums[j] && dp[j]+1 == dp[i]) {
                    cnt[i] += cnt[j];
                }
            }
            maxi = Math.max(dp[i], maxi);
        }
        int res = 0;
        for (int i = 0; i < n; i++) {
            if (dp[i] == maxi) {
                res+=cnt[i];
            }
        }

        return res;
    }
}

File: 0673-number-of-longest-increasing-subsequence/0673-number-of-longest-increasing-subsequence.py
class Solution:
    def findNumberOfLIS(self, nums: List[int]) -> int:
        if len(nums) == 0:
            return 0
        
        dp = [1] * len(nums)      
        cnt = [1] * len(nums)     
        
        for i in range(len(nums)):
            for j in range(i):
                if nums[i] > nums[j]:
                    if dp[j] + 1 > dp[i]:
                        dp[i] = dp[j] + 1
                        cnt[i] = cnt[j]
                    elif dp[j] + 1 == dp[i]:
                        cnt[i] += cnt[j]
        
        maxi = max(dp)
        
        res = sum(cnt[i] for i in range(len(nums)) if dp[i] == maxi)
        return res
